File: ios/Sources/UIKit/XTUIContext.m
//
//  XTUIContext.m
//  XTSample
//
//  Created by 崔明辉 on 2017/8/24.
//  Copyright © 2017年 UED Center, YY Inc. All rights reserved.
//

#import "XTUIContext.h"
#import "XTContext.h"
#import "XTComponent.h"
#import "XTUIApplication.h"
#import "XTUIView.h"
#import "XTUIWindow.h"
#import "XTUIApplicationDelegate.h"
#import "XTUIViewController.h"
#import "XTUINavigationBar.h"
#import "XTUINavigationController.h"
#import "XTUIScreen.h"
#import "XTUIImage.h"
#import "XTUIImageView.h"
#import "XTUILabel.h"
#import "XTUILayoutConstraint.h"
#import "XTUIButton.h"
#import "XTUIFont.h"
#import "XTUIUtils.h"
#import "XTUIScrollView.h"
#import "XTUIListView.h"
#import "XTUIListCell.h"
#import "XTUITextField.h"
#import "XTUITextView.h"
#import "XTDebug.h"
#import "XTUICanvasView.h"
#import "XTUIDevice.h"
#import "XTUITextMeasurer.h"
#import "XTUIHRView.h"
#import "XTUIModal.h"
#import "XTUIWebView.h"
#import "XTUISwitch.h"
#import "XTUISlider.h"
#import "XTUIActivityIndicatorView.h"
#import "XTDebug.h"
#import "XTMemoryManager.h"
#import <JavaScriptCore/JavaScriptCore.h>

@interface XTUIContext ()<UINavigationControllerDelegate, XTDebugDelegate>

@property (nonatomic, readwrite) NSURL *sourceURL;
@property (nonatomic, copy) NSDictionary *initialOptions;

@end

@implementation XTUIContext

+ (NSString *)name {
    return @"_XTUIContext";
}

+ (NSString *)xtr_startWithNamed:(NSString *)name options:(JSValue *)options completion:(JSValue *)completion {
    return [self xtr_startWithURL:[NSURL fileURLWithPath:[[NSBundle mainBundle] pathForResource:name ofType:nil]
                                             isDirectory:NO].absoluteString
                          options:options
                       completion:completion
                          failure:nil];
}

+ (NSString *)xtr_startWithURL:(NSString *)URLString
                       options:(JSValue *)options
                    completion:(JSValue *)completion
                       failure:(JSValue *)failure {
    XTUIContext *context = [[XTUIContext alloc] initWithSourceURL:[NSURL URLWithString:URLString]
                                                          options:options.isObject ? options.toDictionary : nil
                                                  completionBlock:^(UIViewController<XTComponent> * _Nullable rootViewController) {
                                                      if ([rootViewController conformsToProtocol:@protocol(XTComponent)]) {
                                                          [completion callWithArguments:@[rootViewController.objectUUID ?: [NSNull null]]];
                                                      }
                                                   } failureBlock:^(NSError * _Nonnull error) {
                                                       
                                                   }];
    XTManagedObject *managedObject = [[XTManagedObject alloc] initWithObject:context];
    context.objectUUID = managedObject.objectUUID;
    [XTMemoryManager add:managedObject];
    return managedObject.objectUUID;
}

#ifdef LOGDEALLOC
- (void)dealloc {
    NSLog(@"XTUIContext dealloc.");
}
#endif

- (instancetype)initWithSourceURL:(NSURL *)sourceURL
                          options:(nullable NSDictionary *)options
                  completionBlock:(XTUIContextCompletionBlock)completionBlock
                     failureBlock:(XTUIContextFailureBlock)failureBlock
{
    self = [super init];
    if (self) {
        _sourceURL = sourceURL;
        _initialOptions = options;
        [[NSOperationQueue mainQueue] addOperationWithBlock:^{
            [self loadViaSourceURL:completionBlock failureBlock:failureBlock];
        }];
    }
    return self;
}

- (void)setup {
    [super setup];
    [self loadComponents];
    [self loadScript];
}

- (void)loadViaSourceURL:(XTUIContextCompletionBlock)completionBlock failureBlock:(XTUIContextFailureBlock)failureBlock {
    if (self.sourceURL.isFileURL) {
        NSString *script = [[NSString alloc] initWithContentsOfURL:self.sourceURL encoding:NSUTF8StringEncoding error:NULL];
        if (script) {
            [self evaluateScript:script];
            [self.application.delegate didFinishLaunchingWithOptions:self.initialOptions application:self.application];
            if (completionBlock) {
                completionBlock(self.application.delegate.window.rootViewController);
            }
        }
    }
    else {
        NSURLRequest *request = [NSURLRequest requestWithURL:self.sourceURL
                                                 cachePolicy:NSURLRequestUseProtocolCachePolicy
                                             timeoutInterval:15.0];
        [[[NSURLSession sharedSession] dataTaskWithRequest:request
                                         completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
                                             if (error == nil && data != nil) {
                                                 NSString *script = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
                                                 if (script) {
                                                     [[NSOperationQueue mainQueue] addOperationWithBlock:^{
                                                         [self evaluateScript:script];
                                                         [self.application.delegate didFinishLaunchingWithOptions:self.initialOptions application:self.application];
                                                         if (completionBlock) {
                                                             completionBlock(self.application.delegate.window.rootViewController);
                                                         }
                                                     }];
                                                 }
                                             }
                                             else {
                                                 [[NSOperationQueue mainQueue] addOperationWithBlock:^{
                                                     if (failureBlock) {
                                                         failureBlock(error ?: [NSError errorWithDomain:@"XTUIContext" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"unknown error."}]);
                                                     }
                                                     
                                                 }];
                                             }
                                         }] resume];
    }
}

- (void)loadComponents {
    for (Class component in @[
                              [XTUIContext class],
                              [XTUIApplication class],
                              [XTUIApplicationDelegate class],
                              [XTUIView class],
                              [XTUIWindow class],
                              [XTUIViewController class],
                              [XTUINavigationBar class],
                              [XTUINavigationController class],
                              [XTUIScreen class],
                              [XTUIImage class],
                              [XTUIImageView class],
                              [XTUILabel class],
                              [XTUILayoutConstraint class],
                              [XTUIButton class],
                              [XTUIFont class],
                              [XTUIScrollView class],
                              [XTUIListView class],
                              [XTUIListCell class],
                              [XTUITextField class],
                              [XTUITextView class],
                              [XTUICanvasView class],
                              [XTUIDevice class],
                              [XTUITextMeasurer class],
                              [XTUIHRView class],
                              [XTUIModal class],
                              [XTUIWebView class],
                              [XTUISwitch class],
                              [XTUISlider class],
                              [XTUIActivityIndicatorView class],
                              [XTDebug class],
                              ]) {
        if ([component conformsToProtocol:@protocol(XTComponent)]) {
            self[[component name]] = component;
        }
    }
}

- (void)loadScript {
    NSString *path = [[NSBundle mainBundle] pathForResource:@"xt.uikit.ios.min" ofType:@"js"];
    if (path) {
        NSString *script = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:NULL];
        [self evaluateScript:script];
    }
}

#pragma mark - Start Application Static Methods

static NSArray *defaultAttachContextClasses;

+ (void)addDefaultAttachContext:(Class)attachContextClass {
    if ([defaultAttachContextClasses containsObject:attachContextClass]) {
        return;
    }
    if ([attachContextClass isSubclassOfClass:[XTContext class]]) {
        NSMutableArray *classes = (defaultAttachContextClasses ?: @[]).mutableCopy;
        [classes addObject:NSStringFromClass(attachContextClass)];
        defaultAttachContextClasses = [classes copy];
    }
}

static UINavigationController *currentDebugNavigationViewController;
static XTUIContext *currentDebugContext;

+ (void)debugWithIP:(NSString *)IP port:(NSInteger)port navigationController:(UINavigationController *)navigationController {
    [[XTDebug sharedDebugger] setDelegate:(id)self];
    [XTDebug debugWithIP:IP port:port navigationController:navigationController];
    currentDebugNavigationViewController = navigationController;
}

+ (void)debuggerDidTerminal {
    UIViewController *targetViewController = nil;
    BOOL found = NO;
    for (id item in [currentDebugNavigationViewController childViewControllers]) {
        if ([item isKindOfClass:[XTUIViewController class]]) {
            found = YES;
            break;
        }
        targetViewController = item;
    }
    if (found) {
        if (targetViewController == nil) {
            [currentDebugNavigationViewController setViewControllers:@[] animated:NO];
        }
        else {
            [currentDebugNavigationViewController popToViewController:targetViewController animated:NO];
        }
    }
}

+ (void)debuggerDidReload {
    [self debuggerDidTerminal];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.3 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
        currentDebugContext = [[XTUIContext alloc] initWithSourceURL:[[XTDebug sharedDebugger] sourceURL]
                                                             options:@{@"onDebug": @(YES)}
                                                     completionBlock:^(UIViewController * _Nullable rootViewController) {
                                                         [currentDebugNavigationViewController pushViewController:rootViewController animated:YES];
                                                     } failureBlock:nil];
        [[XTDebug sharedDebugger] setDebugContext:currentDebugContext];
    });
}

+ (NSString *)debuggerEval:(NSString *)code {
    if (currentDebugContext != nil) {
        return [currentDebugContext evaluateScript:code].toString;
    }
    return nil;
}

- (void)debuggerDidReload {}
- (void)debuggerDidTerminal {}
- (NSString *)debuggerEval:(NSString *)code { return nil; }

@end

File: android/uikit/src/main/java/com/opensource/xt/uikit/XTUIContext.kt
package com.opensource.xt.uikit

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Handler
import android.util.Log
import com.eclipsesource.v8.V8Function
import com.eclipsesource.v8.V8Object
import com.eclipsesource.v8.utils.V8ObjectUtils
import com.opensource.xt.core.*
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.net.URI
import kotlin.system.measureTimeMillis

/**
 * Created by cuiminghui on 2017/8/31.
 */
class XTUIContext(appContext: android.content.Context,
                  val sourceURL: String?,
                  val options: Map<String, Any>?,
                  val completionBlock: ((bridge: XTUIContext) -> Unit)? = null,
                  val failureBlock: ((e: Exception) -> Unit)? = null): XTContext(appContext), XTComponentInstance {

    companion object: XTDebugDelegate {

        private var defaultAttachContext: MutableList<Class<XTContext>> = mutableListOf()

        fun addDefaultAttachContext(attachContextClass: Class<XTContext>) {
            if (XTContext::class.java.isAssignableFrom(attachContextClass) && !defaultAttachContext.contains(attachContextClass)) {
                defaultAttachContext.add(attachContextClass)
            }
        }

        fun createWithAssets(appContext: android.content.Context,
                             assetsName: String,
                             options: Map<String, Any>?,
                             completionBlock: ((uiContext: XTUIContext) -> Unit)? = null): XTUIContext {
            return createWithSourceURL(appContext, "file:///android_asset/$assetsName", options, completionBlock)
        }

        fun createWithSourceURL(appContext: android.content.Context,
                                sourceURL: String?,
                                options: Map<String, Any>?,
                                completionBlock: ((uiContext: XTUIContext) -> Unit)? = null,
                                failureBlock: ((e: Exception) -> Unit)? = null): XTUIContext {
            val context = XTUIContext(appContext, sourceURL, options, completionBlock, failureBlock)
            defaultAttachContext.forEach {
                try {
                    it.getDeclaredConstructor(android.content.Context::class.java, XTContext::class.java)
                            .newInstance(appContext, context)
                } catch (e: Exception) { e.printStackTrace() }
            }
            return context
        }

        var currentDebugApplicationContext: Context? = null
        var currentDebugContext: XTUIContext? = null

        override fun debuggerDidTerminal() {
            currentDebugContext?.let { currentDebugContext ->
                currentDebugContext.sharedHandler.post {
                    currentDebugContext.release()
                    this.currentDebugContext = null
                }
            }
        }

        override fun debuggerDidReload() {
            currentDebugApplicationContext?.let { currentDebugApplicationContext ->
                Handler(currentDebugApplicationContext.mainLooper).post {
                    XTDebug.sharedDebugger.sourceURL?.let {
                        currentDebugContext = XTUIContext.createWithSourceURL(currentDebugApplicationContext, it, mapOf(Pair("onDebug", true)), {
                            it.start()
                        })
                        XTDebug.sharedDebugger.debugContext = currentDebugContext
                    }
                }
            }
        }

        override fun debuggerEval(code: String, callback: (value: String) -> Unit) {
            currentDebugContext?.let { currentDebugContext ->
                currentDebugContext.sharedHandler.post {
                    try {
                        callback(currentDebugContext.evaluateScript(code) as? String ?: "")
                    } catch (e: Exception) { callback(e.message ?: "") }
                }
            }
        }

    }

    class JSExports(val context: XTUIContext): XTComponentExport() {

        override val name: String = "_XTUIContext"

        override fun exports(): V8Object {
            val exports = V8Object(context.runtime)
            exports.registerJavaMethod(this, "xtr_startWithNamed", "xtr_startWithNamed", arrayOf(String::class.java, V8Object::class.java, V8Function::class.java))
            exports.registerJavaMethod(this, "xtr_startWithURL", "xtr_startWithURL", arrayOf(String::class.java, V8Object::class.java, V8Function::class.java, V8Function::class.java))
            return exports
        }

        fun xtr_startWithNamed(name: String, options: V8Object, completion: V8Function): String {
            val completion = completion.twin()
            val createOptions = mutableMapOf<String, Any>()
            try {
                V8ObjectUtils.toMap(options).forEach {
                    it.value?.let { value ->
                        createOptions.put(it.key, value)
                    }
                }
            } catch (e: Exception) {}
            val context = XTUIContext.createWithAssets(context.appContext, name, createOptions.toMap(), {
                XTContext.callWithArgument(completion, it.application?.delegate?.window?.rootViewController?.objectUUID)
                XTContext.release(completion)
            })
            val managedObject = XTManagedObject(context)
            context.objectUUID = managedObject.objectUUID
            XTMemoryManager.add(managedObject)
            return managedObject.objectUUID
        }

        fun xtr_startWithURL(URLString: String, options: V8Object, completion: V8Function, failure: V8Function): String {
            val completion = completion.twin()
            val failure = failure.twin()
            val createOptions = mutableMapOf<String, Any>()
            try {
                V8ObjectUtils.toMap(options).forEach {
                    it.value?.let { value ->
                        createOptions.put(it.key, value)
                    }
                }
            } catch (e: Exception) {}
            val context = XTUIContext.createWithSourceURL(context.appContext, URLString, createOptions, {
                XTContext.callWithArgument(completion, it.application?.delegate?.window?.rootViewController?.objectUUID)
                XTContext.release(completion, failure)
            }, {
                XTContext.callWithArgument(failure, it.message ?: "")
                XTContext.release(completion, failure)
            })
            val managedObject = XTManagedObject(context)
            context.objectUUID = managedObject.objectUUID
            XTMemoryManager.add(managedObject)
            return managedObject.objectUUID
        }

    }

    var application: XTUIApplication? = null

    override var objectUUID: String? = null

    private var isUIContextDidSetup = false

    init {
        Handler().post({
            loadViaSourceURL()
        })
    }

    fun start() {
        val intent = Intent(appContext, XTUIActivity::class.java)
        intent.putExtra("XTUIShowBackButton", true)
        this.application?.delegate?.window?.rootViewController?.objectUUID?.let {
            intent.putExtra("ViewControllerObjectUUID", it)
        }
        appContext.startActivity(intent)
    }

    fun attach(activity: Activity) {
        this.application?.delegate?.window?.rootViewController?.setContentView(activity)
    }

    fun attach(activity: Activity, fragmentID: Int) {
        this.application?.delegate?.window?.rootViewController?.attachFragment(activity, fragmentID)
    }

    override fun setup() {
        super.setup()
        if (!isUIContextDidSetup) {
            loadComponents()
            loadScript()
            isUIContextDidSetup = true
        }
    }

    private fun loadComponents() {
        val components: List<XTComponentExport> = listOf(
                XTUIContext.JSExports(this),
                XTUIImage.JSExports(this),
                XTUIApplication.JSExports(this),
                XTUIApplicationDelegate.JSExports(this),
                XTUIView.JSExports(this),
                XTUIWindow.JSExports(this),
                XTUIViewController.JSExports(this),
                XTUIButton.JSExports(this),
                XTUIImageView.JSExports(this),
                XTUILabel.JSExports(this),
                XTUICanvasView.JSExports(this),
                XTUITextField.JSExports(this),
                XTUITextView.JSExports(this),
                XTUIDevice.JSExports(this),
                XTUIScreen.JSExports(this),
                XTUINavigationController.JSExports(this),
                XTUIScrollView.JSExports(this),
                XTUITextMeasurer.JSExports(this),
                XTUIFont.JSExports(this),
                XTUIHRView.JSExports(this),
                XTUINavigationBar.JSExports(this),
                XTUIModal.JSExports(this),
                XTUIWebView.JSExports(this),
                XTUISwitch.JSExports(this),
                XTUISlider.JSExports(this),
                XTUIActivityIndicatorView.JSExports(this)
        )
        components.forEach {
            val obj = it.exports()
            this.runtime.add(it.name, obj)
            XTContext.release(obj)
            _registeredComponents.put(it.name, it)
        }
    }

    private fun loadScript() {
        this.appContext.assets.open("xt.uikit.android.min.js")?.let {
            val byteArray = ByteArray(it.available())
            it.read(byteArray)
            String(byteArray).let { this.evaluateScript(it) }
            it.close()
        }
    }

    private fun loadViaSourceURL() {
        val handler = Handler()
        sourceURL?.let { sourceURL ->
            if (sourceURL.startsWith("file://")) {
                try {
                    val inputStream = if (sourceURL.startsWith("file:///android_asset/")) appContext.assets.open(sourceURL.replace("file:///android_asset/", "")) else File(URI.create(sourceURL)).inputStream()
                    val byteArray = ByteArray(inputStream.available())
                    inputStream.read(byteArray)
                    inputStream.close()
                    val script = String(byteArray)
                    handler.post {
                        this.evaluateScript(script)
                        application?.delegate?.didFinishLaunchWithOptions(options ?: mapOf(), application!!)
                        completionBlock?.invoke(this)
                    }
                } catch (e: java.lang.Exception) { handler.post { failureBlock?.invoke(e) };  e.printStackTrace() }
            }
            else if (sourceURL.startsWith("http://") || sourceURL.startsWith("https://")) {
                Thread(Thread.currentThread().threadGroup, {
                    try {
                        val req = Request.Builder().url(sourceURL).method("GET", null).build()
                        val res = OkHttpClient().newCall(req).execute()
                        val script = res.body()?.string() ?: return@Thread
                        handler.post {
                            this.evaluateScript(script)
                            application?.delegate?.didFinishLaunchWithOptions(options ?: mapOf(), application!!)
                            completionBlock?.invoke(this)
                        }
                    } catch (e: Exception) { handler.post { failureBlock?.invoke(e) }; e.printStackTrace() }
                }, "XTUIEval").start()
            }
            else {
                Log.w("XTUIContext", "Unknown sourceURL type >>> $sourceURL")
            }
            return
        }
    }

    override fun release() {
        super.release()
        (application?.delegate?.window?.rootViewController as? XTUINavigationController)?.let {
            it.childViewControllers?.forEach {
                it.activity?.finish()
            }
        }
        application?.delegate?.window?.rootViewController?.activity?.finish()
    }

}